### DISCUSSION TITLE
gnssntripclient running for around 60 minutes only

### DISCUSSION THREAD

**Discussion by leoheck** [semuconsulting/pygnssutils#45] [General]
Hi, I am using `gnssmqttclient 1.0.16` to monitor messages a caster gives me. This caster is receiving messages from my forwarder. But the is not running more than 60 minutes. 

I have increased the default waittime from 3 to 10. 

Is there any other thing that I can do to make the `gnssmqttclient` run longer?

**Comment by semuadmin** [comment]
Hi @leoheck

Can you describe your configuration? `gnssmqttclient` is designed to connect directly to a SPARTN IP (MQTT) server, such as `pp.services.u-blox.com`. Which MQTT server are you using? What do you mean by 'forwarder'?

When you say it's "not running more than 60 minutes", what's actually happening? Is it terminating after 60 minutes or simply not receiving any more SPARTN data from the MQTT server? If it's terminating, are you getting an error message?

Could you run it with the following command line options and send me the log file please?

`--verbosity 2 --logtofile 1`

**Comment by leoheck** [comment]
Oh, @semuadmin, sorry, I mentioned the wrong tool when I went to get the version. I meant `gnssntripclient`.

I am running it like this:

```bash
gnssntripclient \
	--ntripuser ${user} \
	--ntrippassword ${password} \
	--waittime 10 \
	-S ${caster_addr} \
	-P 2101 \
	-M ${mountpoint} > "./ntrip_log_${mountpoint}.csv"
```

I am also piping it to a script to format the output a bit better for my gnuplot script. But it does not run longer than let's say 60 minutes. I have one `ntrip-client` but it is not that good since it may be losing message and I am trying to use the `gnssntripclient`, instead.

Does it have a time limit or something?

Edit: Thanks for the quick reply.

**Reply by leoheck** [reply]
I am not using the `--logfile 1` since I cannot change the `file_path` only the `folder_path`. It could be the `file_path` instead where I can give a path and also name the file. But this is not the problem.

I am receiving data.. the issue is that it does not run long enough. I would like to collect data for a couple of days to evaluate the availability of my system. 

I am running it on a Raspberry Pi 3, by the way.

**Comment by semuadmin** [comment]
If you can answer these questions in full I can endeavour to help:

1. Can you provide the COMPLETE, ACTUAL Python script that you're running please (_you can redact the userid and password_)?
1. Can you describe your complete hardware and software configuration? What do you mean by 'forwarder'?
1. Which NTRIP caster and mountpoint are you using? Do you have/need a registered account with this service?
1. Have you checked whether the caster/mountpoint you are using requires you to send NMEA GGA client position sentences? Many NTRIP casters will stop sending data if they do not receive such GGA sentences - check with your NTRIP service.
1. Is your script terminating after 60 minutes, or simply not receiving any more data from the NTRIP caster?
1. If it's terminating, are you getting any error or information message of any kind?
1. You appear to be piping the output to a CSV file. Are you sure you're not simply running out of space on your RPi's SD card?

gnssntripclient does not implement an arbitrary timelimit, but some NTRIP casters will stop sending data after a period of time - check with your NTRIP service.

**Reply by leoheck** [reply]
I will try to address your questions inline.

> Can you provide the COMPLETE, ACTUAL Python script that you're running please (you can redact the userid and password)?

```bash
gnssntripclient \
	--ntripuser ${user} \
	--ntrippassword ${password} \
	--waittime 10 \
	-S ${caster_addr} \
	-P 2101 \
	-M ${mountpoint} > "./ntrip_log_${mountpoint}.csv"
```

This is inside of a systemd service so if the command breaks or something happens it starts again after a couple of seconds.


> Can you describe your complete hardware and software configuration? What do you mean by 'forwarder'?

Ah, my bad I call forwarder my `ntripserver`.

The hardware of the client is an `Raspberry Pi 3B+ or 4B` or on my computer `Dell XPS 9550`
The server is also another `Raspberry Pi 3B+ or 4B`.

> Which NTRIP caster and mountpoint are you using? Do you have/need a registered account with this service?

I am using Ublox Caster, I need a registered account. I am not sure if I can share the mountpoint.

Have you checked whether the caster/mountpoint you are using requires you to send NMEA GGA client position sentences? Many NTRIP casters will stop sending data if they do not receive such GGA sentences - check with your NTRIP service.

Oh, that's an interesting statement. I am not sure but this seems to be important. Now I understand why the client has these options below. Maybe I can tweak the client a bit with these settings.

```
  --ggainterval GGAINTERVAL
                        GGA sentence transmission interval (-1 = None) (default: -1)
  --ggamode {0,1}       GGA pos source; 0 = live from receiver, 1 = fixed reference (default: 0)
  --reflat REFLAT       reference latitude (default: 0.0)
  --reflon REFLON       reference longitude (default: 0.0)
  --refalt REFALT       reference altitude (default: 0.0)
  --refsep REFSEP       reference separation (default: 0.0)
```

> Is your script terminating after 60 minutes, or simply not receiving any more data from the NTRIP caster?

The data reception is stopping for sure. Now I am not sure if the `gnssntripclient` is terminating anymore but if that was happening I would have a reconnect and maybe I would see fresh data comming again. 

> If it's terminating, are you getting any error or information message of any kind?

This is why I think it is not terminating. Otherwise, I would have shared the error in my first message here. But I am not 100% sure anymore. I will have to run another test.

> You appear to be piping the output to a CSV file. Are you sure you're not simply running out of space on your RPi's SD card?

Yeah, that's not the problem. I have at least 64 GB available and when I pipe it to a file just a few columns of the output. Not the whole RTCM3 parsed message. It is like the datetime and the data until the first column where I can see the `msg_id` which is the first thing in the parsed RTCM message.

> gnssntripclient does not implement an arbitrary timelimit, but some NTRIP casters will stop sending data after a period of time - check with your NTRIP service.

I don't think this is the issue but I understand the caster I am using may restart/stop after some time.

**Comment by leoheck** [comment]
Now you mention it, what happens with the gnssntripclient if the caster stops sending? Does it remain connected to listening or does it exist? I am asking since I believe the client is waiting forever but without receiving any extra message and it does not try to reconnect again automatically.

**Comment by semuadmin** [comment]
Ok you appear to have a lot of moving parts in your configuration which I'm only partially sighted on, so I can only offer broad advice.

Like the other `pygnssutils` tools, `gnssntripclient` was really only intended to be used as a simple CLI diagnostic tool - it's not in itself intended to be used as production NTRIP client and certainly not as part of some unattended systemd process.

Once connected, `gnssntripclient` will run indefinitely until explicitly stopped by the user, _regardless_ of whether it is receiving NTRIP data or not. There is no 'inactivity timeout' or 'retry' as such, unless you choose to implement one yourself in your application code. If the NTRIP caster has closed the TCP socket connection (or it's dropped for any other reason e.g. internet connectivity issue) and you're sending GGA sentences periodically from `gnssntripclient` (`--ggainterval > 0`), you would get a 'socket error' on the next periodic send GGA attempt, but otherwise it will simply remain quietly connected.

FYI the sourcetable (STR) entry for your particular mountpoint (_information which is generally in the public domain, by the way_) gives you lots of information about the mountpoint - in particular, field 12 indicates whether an NMEA GGA sentence is required

<img width="874" alt="Screenshot 2023-10-27 at 13 06 39" src="https://github.com/semuconsulting/pygnssutils/assets/28569967/dccce117-00ba-49f2-8584-82bed427c4a7">

If this field is set to '1', you will need to send GGA position date to the caster in order to receive NTRIP correction data back.
